keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+the Application, that match the Application's `Spec.Selector`.
+1. For each selected object, validate that the object has a `kubernetes.io/application` that matches the name of the 
+Application. Discard any objects that do not have such an annotation. This ensures that selector overlap does not 
+violate the intentions of the application creators and administrators.
+1. Unless otherwise indicated, clients should ensure that the `OwnerReferences` of the object contain the `UID` of the 
+Application.
+
+#### Retrieving Applications
+In order to retrieve the Application corresponding to an API object, a client should perform the following.
+
+1. Examine the object for a `kubernetes.io/application` annotation.
+1. If no annotation is found, the object is not associated with an Application.
+1. If the annotation is found, retrieve the Application object with the indicated name from the same namespace of the 
+object.
+1. If no such Application is found, the Application is either deleted or has not yet been created.
+1. If the Application is found, validate that the `Spec.Selector` of the Application object matches the labels of the 
+object. If the object's labels do not match, the object has been orphaned.
+1. Check the `OwnerReferences` of the object for a reference to the Application by `UID` to ensure that ownership has 
+been established.
+
+#### Retrieving Dependencies
+In order to retrieve the `Spec.Dependencies` of an Application, a client should perform the following.
+
+1. For each name in the `Spec.Dependencies` list, retrieve the Application of that name, in the namespace of the 
+Application under consideration. If no such Application is found, the dependency is not installed.
+
+#### Health Checks
+The semantics of readiness for an Application vary too greatly for declarative specification. We propose to encode 
+the readiness semantics of Applications in the same manner as Pods (i.e. by using programs). An ApplicationHealthCheck 
+object contains the template for a Pod that encodes a semantic readiness check for the components of an Application. 
+The requirements of implementing a ApplicationHealthCheck Pod are as follows.
+
+1. A health check Pod must implement a readiness check.
+1. The readiness check of the Pod may only return successfully when the conditions of readiness are met for the entire 
+application. 
+1. An ApplicationHealthCheck Pod should not supply other application telemetry. When its readiness check fails that 
+telemetry would be unavailable (at precisely the time it is most necessary).
+
+### Application Controller
+The Application Controller will be responsible for configuring ownership to enable GC (Garbage Collection) for an 
+Application's components and for updating the `Status` of the Application with respect to installation 
+status of the components and dependencies of the Application and to the readiness of the Application's health check.
```

Comment:
An upgrade of an application handles does what? Are there any operations on an Application that will result in modification to any of the resources beneath it? If a new object is added during an upgrade, what happens? What happens if an object is _removed_ from an application? The GC explanation below leads me to assume that it will be deleted, not merely orphaned. But that's a guess.

Also, is there a way to express that when an application is updated (or one part of an application is updated), some or all of the pods attached need to be restarted/rolled/etc? For example, if I remove a Dependency on mysql, and then add a MySQL deployment to this Application and update the application, is it then up to me to go manually restart all of the parts of my application that are impacted by swapping out the DB?